### PR TITLE
8337506: Disable "best-fit" mapping on Windows command line

### PR DESCRIPTION
**Pull request by naotoj**
Fixing the Java launcher's command line argument parsing issue on Windows. The Java launcher on Windows has been using `GetCommandLineA()` to obtain arguments, which by default does "best-fit" mapping when the arguments are converted to ANSI code page encoding. By disabling this "best-fit" mapping, the launcher's parsing works as expected. A corresponding CSR has been drafted for the behavioral change.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8337509](https://bugs.openjdk.org/browse/JDK-8337509) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8337506](https://bugs.openjdk.org/browse/JDK-8337506): Disable "best-fit" mapping on Windows command line (**Bug** - P4)
 * [JDK-8337509](https://bugs.openjdk.org/browse/JDK-8337509): Disable "best-fit" mapping on Windows command line (**CSR**)

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20428/head:pull/20428` \
`$ git checkout pull/20428`

Update a local copy of the PR: \
`$ git checkout pull/20428` \
`$ git pull https://git.openjdk.org/jdk.git pull/20428/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20428`

View PR using the GUI difftool: \
`$ git pr show -t 20428`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20428.diff">https://git.openjdk.org/jdk/pull/20428.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20428#issuecomment-2263587344)

### CODE REVIEW

File: src/java.base/share/native/launcher/main.c [hunk 1]
```diff
@@ -110,7 +110,23 @@ main(int argc, char **argv)
...
-    JLI_CmdToArgs(GetCommandLine());
+
+    // Obtain the command line in UTF-16, then convert it to ANSI code page
+    // without the "best-fit" option
+    LPWSTR wcCmdline = GetCommandLineW();
+    int mbSize = WideCharToMultiByte(CP_ACP,
+        WC_NO_BEST_FIT_CHARS | WC_COMPOSITECHECK | WC_DEFAULTCHAR,
+        wcCmdline, -1, NULL, 0, NULL, NULL);
+    LPSTR mbCmdline = JLI_MemAlloc(mbSize);
+    if (WideCharToMultiByte(CP_ACP, WC_NO_BEST_FIT_CHARS | WC_COMPOSITECHECK | WC_DEFAULTCHAR,
```
Comment (AlanBateman) [inline src/java.base/share/native/launcher/main.c:121]: I think this looks okay but will need to handle GetCommandLineW, WideCharToMultiByte, or JLI_MemAlloc failing.

Comment (naotoj) [reply src/java.base/share/native/launcher/main.c:121]: IIUC, all errors should be handled with the proposed patch. On error with JLI_MemAlloc and WideCharToMultiByte, the process exits with `exit(1)`. As to `GetCommandLineW()`, there is no description of error in the MS document (https://learn.microsoft.com/en-us/windows/win32/api/processenv/nf-processenv-getcommandlinew), so I suppose no error handling on our side is needed.

Comment (AlanBateman) [reply src/java.base/share/native/launcher/main.c:121]: > IIUC, all errors should be handled with the proposed patch. On error with JLI_MemAlloc and WideCharToMultiByte, the process exits with `exit(1)`. 

I think I'm mostly wondering about WideCharToMultiByte as it returns 0 when it fails.

Comment (naotoj) [reply src/java.base/share/native/launcher/main.c:121]: The first call to `WideCharToMultiByte()` returns the required size for the multibyte buffer. When it returns 0, `exit(1)` is issued inside `JLI_MemAlloc()`. If the second call to WideCharToMultiByte()` returns 0, it is explicitly dealt with the patch which also issues `exit(1)`.

Comment (AlanBateman) [reply src/java.base/share/native/launcher/main.c:121]: > The first call to `WideCharToMultiByte()` returns the required size for the multibyte buffer. When it returns 0, `exit(1)` is issued inside `JLI_MemAlloc()`.

That's a bit subtle. I think it needs to be handled here or at least include a comment to say that it returns 0 on error which will cause JLI_MemAlloc to exit.

Comment (naotoj) [reply src/java.base/share/native/launcher/main.c:123]: OK, added the comment

**Review by AlanBateman — APPROVED** [review]

**Pull request closed without merging**